r-session-04.qmd

```diff
@@ -0,0 +1,800 @@
+#| message: false
+#| out-width: 100%
+#| column: body
+
+###############################################################################
+# Simulate over time                                                          #
+###############################################################################
+counter <- 0                # set counter at 0
+while(all(I>0)){            #continue until I is depleted
+  counter <- counter + 1    #counter for number of transitions: we don't know how many transitions will happen until the simualtion is over
+  # 
+  SIRtmp <- c(S[counter], I[counter], R[counter], time[counter])       #current SIR states
+  step <- gilstep(SIRtmp, beta, gamma)
+  if(step$change_state ==1){    # if transition is an infection, reduce S and increase I
+    SIRtmp[1] <- SIRtmp[1] - 1
+    SIRtmp[2] <- SIRtmp[2] + 1 
+  }
+  if(step$change_state ==2){    # if transition is an recovery, reduce I and increase R
+    SIRtmp[2] <- SIRtmp[2] - 1
+    SIRtmp[3] <- SIRtmp[3] + 1 
+  }
+  SIRtmp[4] <- SIRtmp[4] + step$timestep      # increment time
+    
+#Append changes
+  S <- c(S,SIRtmp[1])
+  I <- c(I,SIRtmp[2])
+  R <- c(R,SIRtmp[3])
+  time <- c(time,SIRtmp[4])
+
+  #cat(S[counter],"-",I[counter],"-",R[counter],"-",time[counter], ".\n"). # this prints the output as it goes
+}
+```
+
+We can now plot the one realization of this stochastic outbreak. Notice that it has the same general shape as the deterministic outbreak, but is no longer smooth because each individual event over time happens randomly. 
+
+```{r}
+#| echo: true
+#| message: false
+#| out-width: 100%
+#| column: body
+
+# Reuse the plotting code from above
+
+# Turn the output from gilstep() into a tibble (dataframe)
+# so we can manipulate and plot it easily
+gil_df <- tibble(time, S, I, R) %>%  # put elements in a data frame 
+  # Convert all columns to numeric (they are currently type
+  # deSolve so will produce warnings when plotting etc)
+  mutate(
+    # Rather than repeatedly type the same function for every
+    # column, use the across() function to apply the function
+    # to a selection of columns
+    across(
+      # The cols argument takes a selection of columns to apply
+      # a function to. Here, we want to apply the as.numeric()
+      # function to all columns, so we use the function
+      # everything() to select all columns.
+      .cols = everything(),
+      .fns = as.numeric
+    )
+  ) %>%
+  # Convert the dataframe from wide to long format, so we have a
+  # column for the time, a column for the state, and a column
+  # for the proportion of the population in that state at that
+  # time
+  pivot_longer(
+    # Don't pivot the time column
+    cols = -c(time),
+    names_to = "state",
+    values_to = "number"
+  ) %>%
+  # Update the state column to be a factor, so the plot will
+  # show the states in the correct order
+  mutate(state = factor(state, levels = c("S", "I", "R")))
+
+SIRcolors <- c(S = "#1f77b4", I = "#ff7f0e", R = "#FF3851")
```

Comment:
It should be ok to remove the redefinition of colors as the variable exists in previous chunks